driver/scsi_operation.c

```diff
@@ -410,9 +530,75 @@ WnbdPendOperation(_In_ PVOID DeviceExtension,
             break;
         }
         Status = WnbdPendElement(DeviceExtension, ScsiDeviceExtension, Srb,
-            BlockAddress * ScsiInfo->UserEntry->BlockSize, DataLength);
+            BlockAddress * ScsiInfo->UserEntry->BlockSize, DataLength, (BOOLEAN)FUA);
         }
         break;
+    case SCSIOP_UNMAP:
+        PUNMAP_LIST_HEADER DataBuffer = SrbGetDataBuffer(Srb);
+        ULONG DataTransferLength = SrbGetDataTransferLength(Srb);
+        UINT16 BlockDescLength;
+
+        if (!DataBuffer ||
+            DataTransferLength < sizeof(UNMAP_LIST_HEADER) ||
+            DataTransferLength < sizeof(UNMAP_LIST_HEADER) + (
+                BlockDescLength = ((ULONG)DataBuffer->BlockDescrDataLength[0] << 8) |
+                                   (ULONG)DataBuffer->BlockDescrDataLength[1]) ||
+            BlockDescLength > WNBD_MAX_TRANSFER_LENGTH)
+        {
+            Srb->SrbStatus = SRB_STATUS_ABORTED;
+            Status = STATUS_BUFFER_TOO_SMALL;
+            break;
+        }
+
+        // We don't support the anchored state at the moment.
+        if (Cdb->UNMAP.Anchor)
+        {
+            WNBD_LOG_ERROR("Unmap 'anchor' state not supported.");
+            Srb->SrbStatus = SRB_STATUS_INVALID_REQUEST;
+            Status = STATUS_INVALID_PARAMETER;
```

Comment:
Storport might not be notified unless we call StorPortNotification here. There are a few other similar cases.